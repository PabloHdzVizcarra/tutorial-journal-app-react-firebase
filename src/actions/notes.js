import { db } from "../firebase/firebase-config";
import {types} from '../types/types';
import { loadNotes } from "../helpers/loadNotes";
import Swal from 'sweetalert2'
import { fileUpload } from "../helpers/fileUpload";



// creando nuevas notas
export const startNewNote = () => {

  // podemos obtener todo el state con la funcion getState gracias a reduxThunk
  return async (dispatch, getState) => {

    // obtenemos el id del usuario
    const { uid } = getState().auth;
    
    const newNote = {
      title: '',
      body: '',
      date: new Date().getTime()
    }

    try {
      
      // agregamos la nota a firestore
      const doc = await db.collection(`${uid}/journal/notes`).add(newNote);

      // se agrega la nota al state active
      dispatch(activeNote(doc.id, newNote));
  
      // modificamos el state con la nueva nota
      dispatch(addNewNote(doc.id, newNote));

    } catch (error) {
      console.log(error);
      
    }


  }
}

export const activeNote = (id, note) => ({
  type: types.notesActive,
  payload: {
    id,
    ...note
  }
});

export const addNewNote = (id, note) => ({
  type: types.notesAddNew,
  payload: {
    id,
    ...note
  }
});

export const startLoadingNotes = (uid) => {
  return async (dispatch) => {

    // se obtienen las notas mediante el helper
    const notes = await loadNotes(uid);

    // se cargan las notes al state mediante el helper y action y reducer
    dispatch(setNotes(notes));
  } 
}

export const setNotes = (notes) => ({
  type: types.notesLoad,
  payload: notes
})

export const startSaveNote = (note) => {
  return async (dispatch, getState) => {

    // se obtiene el acceso por el id unico del usuario
    const { uid } = getState().auth;

    // comprobacion de la propiedad url de note
    if (!note.url) {
      delete note.url;
    }

    // objeto creado con spread con todos los datos
    const noteToFirestore = { ...note };

    // elimina el id del objeto
    delete noteToFirestore.id;

    // se obtiene la referencia al documento especificando la ruta
    await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore);

    // se ejecuta un dispatch para refrescar las notas del sidebar
    dispatch(refreshNote(note.id, noteToFirestore));
    Swal.fire('Nota guardada', note.title, 'success');
  }
}

export const refreshNote = (id, note) => ({

  type: types.notesUpdate,
  payload: {
    id,
    note: {
      id, 
      ...note
    }
  }
})

export const startUploading = (file) => {
  return async (dispatch, getState) => {

    const { active: activeNote } = getState().notes;

    // alerta cuando empieza a cargar la imagen
    Swal.fire({
      title: 'Cargando...',
      text: 'Espere un momento...',
      allowOutsideClick: false,
      onBeforeOpen: () => {
        Swal.showLoading();
      }
    })

    // helper para obtener el link de la imagen subida
    const fileUrl = await fileUpload(file);

    // agregando el url a la nota activa
    activeNote.url = fileUrl;
    
    // grabando la nota 
    dispatch(startSaveNote(activeNote));

    // se cierra la alerta
    Swal.close();
  }
}

export const startDeleting = (id) => {
  return async (dispatch, getState) => {

    // obtener state de redux
    const uid = getState().auth.uid;

    // eliminando la nota de firebase con su url
    await db.doc(`${uid}/journal/notes/${id}`).delete();

    // se ejecuta la action y se envia al reducer
    dispatch(deleteNote(id));

  }
}

export const deleteNote = (id) => ({
  type: types.notesDelete,
  payload: id
});

// action
export const noteLogout = () => ({
  type: types.notesLogoutCleaning
  
});