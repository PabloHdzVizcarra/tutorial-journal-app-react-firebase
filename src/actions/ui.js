import { types } from "../types/types";


export const setError = (error) => ({

  type: types.uiSetError,
  payload: error
});

export const removeError = () => ({

  type: types.uiRemoveError,
});

// action para inidicar la carga
export const startLoading = () => ({
  type: types.uiStartLoading,
    
})

// action para inidicar la carga
export const finishLoading = () => ({
  type: types.uiFinishLoading
    
})