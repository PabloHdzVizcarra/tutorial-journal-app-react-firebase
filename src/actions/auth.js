import Swal from 'sweetalert2';

import { types } from "../types/types";
import {firebase, googleAuthProvider} from '../firebase/firebase-config';
import { startLoading, finishLoading } from "./ui";
import { noteLogout } from './notes';

export const startLoginEmailPassword = (email, password) => {
  // cuando sean tareas asincronas se retorna un callback, a ese callback le tienes que pasar como argumento el disptach a ejecutar
  return (dispatch) => {

    dispatch(startLoading());// verificando si esta cargando

    return firebase
      .auth()
      .signInWithEmailAndPassword(email, password)
      .then((data) => {
        const { user } = data; //destrcuturing data

        

        // ejecutando el dispatch, modificando el state
        dispatch(
          login(
            user.uid,
            user.displayName
          )
        )

        dispatch(finishLoading());// verificando si esta cargando

      })
      .catch((error) => {
        dispatch(finishLoading());// verificando si esta cargando
        console.log(error);
        Swal.fire('Error, ', error.message, 'error');
      })
    
  }
}

export const startRegisterWithEmailPasswordName = (email, password, name) => {
  return (dispatch) => {

    firebase
      .auth()
      .createUserWithEmailAndPassword(email, password)
      .then(async ({ user }) => { // destructuring obteniendo datos del user
        // se usa async/await para manejar las promesas mas eficientemente
        
        // con este metodo de firebase se agrega el nombre de usuario y una foto de perfil si es requerido (este metodo retorna una promesa)
        await user.updateProfile({ displayName: name });

        // dispatch para modificar el state
        dispatch(
          login(
            user.uid,
            user.displayName
          )
        )
      })
      .catch((error) => {
        console.log(error);
        Swal.fire('Error, ', error.message, 'error');
      })
  }
}

export const startGoogleLogin = () => {
  return (dispatch) => {

    // se usa firebase.auth para autenticar ese usuario
    firebase.auth().signInWithPopup(googleAuthProvider)
      .then(({user}) => { // destructuring para obtener el user
        dispatch(
          login( // ejecutando el otro dispatch ya con la data obtenida
            user.uid,
            user.displayName
          )
        )
      })
  }
}

export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName       
  }
});

export const startLogout = () => {
  return async (dispatch) => {

    try {
      // metodo para cerrar la cuenta en firebase
      await firebase.auth().signOut();

      // action para modificar el state y cerrar sesion
      dispatch(logout());

      // action para borrar las notas de redux 
      dispatch(noteLogout());


    } catch (error) {
      console.log(error);
    }
  }
}

export const logout = () => ({
  type: types.logout
});

