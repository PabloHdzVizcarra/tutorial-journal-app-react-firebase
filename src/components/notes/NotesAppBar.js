import React from 'react'
import { startSaveNote, startUploading } from '../../actions/notes'
import { useDispatch, useSelector } from 'react-redux'

export const NotesAppBar = () => {

  const dispatch = useDispatch();
  const { active } = useSelector(state => state.notes);
  

  const handleSave = () => {
    dispatch(startSaveNote(active));
  }

  const handlePictureClicc = () => {

    // simulamos el click al boton de picture
    document.querySelector('#fileSelector').click();
  }

  const handleFileChange = (event) => {

    // obtenemos el archivo que subimos al navegador mediando el evento
    const file = event.target.files[0];

    if (file) {
      dispatch(startUploading(file));
    }
  }

  return (
    <div className="notes__appbar">
      <span>13 de Julio de 2020</span>

      {/* se oculta el input por estetica con display 'none' */}
      <input
        id='fileSelector'  
        type='file'
        name='file'
        style={{display: 'none'}}
        onChange={handleFileChange}
      />

      <div>
        <button
          className="btn"
          onClick={handlePictureClicc}
        >
          Picture
        </button>

        <button
          className="btn"
          onClick={handleSave}
        >
          Save
        </button>
      </div>

    </div>
  )
}
