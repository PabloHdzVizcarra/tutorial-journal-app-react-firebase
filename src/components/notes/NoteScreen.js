import React from 'react'
import { NotesAppBar } from './NotesAppBar'
import { useSelector, useDispatch } from 'react-redux'
import { useForm } from '../../hooks/useForm';
import { useEffect } from 'react';
import { useRef } from 'react';
import { activeNote, startDeleting } from '../../actions/notes';

export const NoteScreen = () => {

  const dispatch = useDispatch();

  const { active: note } = useSelector(state => state.notes);

  // hook personalizado para leer la modificacion de los inputs
  const [formValues, handleInputChange, reset] = useForm(note);

  // destructuring de los valores
  const { body, title, id } = formValues;

  // se crea una variable con useRef para que pueda ser mutable y no renderize el componente
  const activeId = useRef(note.id);

  useEffect(() => {

    if (note.id !== activeId.current) {
      reset(note);
      activeId.current = note.id;
    }
  
  }, [note, reset]);

  useEffect(() => {
    
    // se lee la modificacion cuando cambia el valor de los inputs
    dispatch(activeNote(formValues.id, { ...formValues }));

  }, [formValues, dispatch]);


  // eliminar nota con el boton
  const handleDelete = () => {

    // el id se obtiene de formValues ya que la informacion se establece 
    dispatch(startDeleting(id));
  }

  return (
    <div className="notes__main-content">
      
      <NotesAppBar />

      <div className="notes__content animate__animated animate__fadeIn">
        <input
          type="text"
          name='title'
          placeholder="Some awesome"
          className="notes__title-input"
          autoComplete="off"
          value={title}
          onChange={handleInputChange}
        />

        <textarea
          placeholder="What hapenned today"
          className="notes__textarea"
          name='body'
          value={body}
          onChange={handleInputChange}
        ></textarea>

        {/* Carga condicional de la imagen */}
        {
          (note.url) &&

          <div className="notes__image">
          <img 
            src={note.url}
            alt="Paisaje"
          />
          </div>
        }
      </div>

      <button
        className='btn btn-danger'
        onClick={handleDelete}
      >
        Delete
      </button>
    </div>
  )
}
