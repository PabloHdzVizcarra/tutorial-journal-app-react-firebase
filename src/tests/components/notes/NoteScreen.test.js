import React from 'react';
import '@testing-library/jest-dom';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { NoteScreen } from '../../../components/notes/NoteScreen';
import { activeNote } from '../../../actions/notes';


jest.mock('../../../actions/notes', () => ({
  activeNote: jest.fn()
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initialState = {
  auth: {
    uid: '123',
    name: 'Testing'
  },
  ui: {
    loading: false,
    msgError: null
  },
  notes: {
    notes: [
      
    ],
    active: {
      id: 123,
      title: 'Hello',
      body: 'Mundo',
      date: 17
    }
  }
};

let store = mockStore(initialState);

// simulamos el metodo dispatch de Redux
store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={store}>
      
    <NoteScreen />
    
  </Provider>
);

describe('Test in <NoteScreen />', () => {

  test('should be show correctly ', () => {

    expect(wrapper).toMatchSnapshot();
    
  });


  test('should trigger the active note', () => {
    
    wrapper.find('input[name="title"]').simulate('change', {
      target: {
        name: 'title',
        value: 'Hello Again'
      }
    });

    expect(activeNote).toHaveBeenCalled();

    // con esto verificamos la ultima vez que se llama una funcion
    expect(activeNote).toHaveBeenLastCalledWith(
      123,
      {
        id: 123,
        title: 'Hello Again',
        body: 'Mundo',
        date: 17
      }
    );

  })
  
  
  
})
