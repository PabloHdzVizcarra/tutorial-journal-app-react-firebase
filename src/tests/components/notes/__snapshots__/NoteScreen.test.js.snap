// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <NoteScreen /> should be show correctly  1`] = `
<div
  className="notes__main-content"
>
  <div
    className="notes__appbar"
  >
    <span>
      13 de Julio de 2020
    </span>
    <input
      id="fileSelector"
      name="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content animate__animated animate__fadeIn"
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some awesome"
      type="text"
      value="Hello"
    />
    <textarea
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="What hapenned today"
      value="Mundo"
    />
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
